java/vector/src/main/java/org/apache/arrow/vector/ipc/ArrowStreamWriter.java

```diff
@@ -121,4 +129,45 @@ public static void writeEndOfStream(WriteChannel out, IpcOption option) throws I
   protected void endInternal(WriteChannel out) throws IOException {
     writeEndOfStream(out, option);
   }
+
+  @Override
+  protected void ensureDictionariesWritten(DictionaryProvider provider, Set<Long> dictionaryIdsUsed)
```

Comment:
Oof, so for each batch we write, we need to re-compare all dictionaries? That may also be worth noting in the PR description.

Reply:
That said, this is broadly in line with the C++ implementation: https://github.com/apache/arrow/blob/766e25440250dde9117e19245389badb5ed7678c/cpp/src/arrow/ipc/writer.cc#L1121

The C++ implementation takes a number of shortcuts to avoid having to compare the full data buffers on each batch. However, because vectors are mutable in the Java implementation, we can't easily make the same optimization here.

Reply:
Yeah I don't think there's a way to avoid this due to needing to account for mutable vectors, I've added a comment about this in the PR description.